source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,594 @@
+}
+
+ur_result_t SanitizerInterceptor::enqueueAllocInfo(
+    ur_context_handle_t Context, ur_device_handle_t Device,
+    ur_queue_handle_t Queue, std::shared_ptr<USMAllocInfo> &AllocInfo,
+    ur_event_handle_t &LastEvent) {
+    // Init zero
+    UR_CALL(enqueueMemSetShadow(Context, Device, Queue, AllocInfo->AllocBegin,
+                                AllocInfo->AllocSize, 0, LastEvent,
+                                &LastEvent));
+
+    uptr TailBegin = RoundUpTo(AllocInfo->UserEnd, ASAN_SHADOW_GRANULARITY);
+    uptr TailEnd = AllocInfo->AllocBegin + AllocInfo->AllocSize;
+
+    // User tail
+    if (TailBegin != AllocInfo->UserEnd) {
+        auto Value = AllocInfo->UserEnd -
+                     RoundDownTo(AllocInfo->UserEnd, ASAN_SHADOW_GRANULARITY);
+        UR_CALL(enqueueMemSetShadow(Context, Device, Queue, AllocInfo->UserEnd,
+                                    1, Value, LastEvent, &LastEvent));
+    }
+
+    int ShadowByte;
+    switch (AllocInfo->Type) {
+    case USMMemoryType::HOST:
+        ShadowByte = kUsmHostRedzoneMagic;
+        break;
+    case USMMemoryType::DEVICE:
+        ShadowByte = kUsmDeviceRedzoneMagic;
+        break;
+    case USMMemoryType::SHARE:
+        ShadowByte = kUsmSharedRedzoneMagic;
+        break;
+    case USMMemoryType::MEM_BUFFER:
+        ShadowByte = kMemBufferRedzoneMagic;
+        break;
+    default:
+        ShadowByte = 0xff;
+        assert(false && "Unknow AllocInfo Type");
+    }
+
+    // Left red zone
+    UR_CALL(enqueueMemSetShadow(Context, Device, Queue, AllocInfo->AllocBegin,
+                                AllocInfo->UserBegin - AllocInfo->AllocBegin,
+                                ShadowByte, LastEvent, &LastEvent));
+
+    // Right red zone
+    UR_CALL(enqueueMemSetShadow(Context, Device, Queue, TailBegin,
+                                TailEnd - TailBegin, ShadowByte, LastEvent,
+                                &LastEvent));
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::updateShadowMemory(ur_queue_handle_t Queue) {
+    auto Context = getContext(Queue);
+    auto Device = getDevice(Queue);
+    assert(Device != nullptr);
+
+    auto ContextInfo = getContextInfo(Context);
+
+    auto HostInfo = ContextInfo->getDeviceInfo(nullptr);
+    auto DeviceInfo = ContextInfo->getDeviceInfo(Device);
+    auto QueueInfo = ContextInfo->getQueueInfo(Queue);
+
+    std::shared_lock<ur_shared_mutex> HostGuard(HostInfo->Mutex,
+                                                std::defer_lock);
+    std::unique_lock<ur_shared_mutex> DeviceGuard(DeviceInfo->Mutex,
+                                                  std::defer_lock);
+    std::scoped_lock<std::shared_lock<ur_shared_mutex>,
+                     std::unique_lock<ur_shared_mutex>, ur_mutex>
+        Guard(HostGuard, DeviceGuard, QueueInfo->Mutex);
+
+    ur_event_handle_t LastEvent = QueueInfo->LastEvent;
+
+    // FIXME: Always update host USM, but it'd be better to update host USM
```

Comment:
but this doesn't do anything, no? it will go into `enqueueMemSetShadow`, which will then subsequently die on with unsupported device message.

Can you explain why host memory must be updated every time? won't that overwrite existing data?

Reply:
> but this doesn't do anything, no? it will go into `enqueueMemSetShadow`, which will then subsequently die with unsupported device message.

I don't think so. Here the `Device` is not `nullptr`, as it comes from `Queue`.

> Can you explain why host memory must be updated every time? won't that overwrite existing data?

In fact, the shadow memory of host USM needs to be updated only once per-device, and you can treat these values as read-only data in kernel, and it needn't to be updated until the host USM have been released(I haven't added this logic now).